### PR TITLE
implement pairing for bluetooth proxy

### PR DESCRIPTION
**Pull request by jagheterfredrik**
This patch implements bluetooth pairing for bluetooth proxy on ESP32.

# What does this implement/fix?

The pairing function was not implemented

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):**

 - https://github.com/dbuezas/eq3btsmart/issues/4 
 - https://github.com/bendikwa/esphome-igrill/blob/main/components/igrill/igrill.cpp#L34

## Test Environment

- [ ] ESP32
- [X] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

https://esphome.github.io/bluetooth-proxies/

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by jesserockz** [comment]
While this change seems to be working, it does bring in a bunch of error logs from esp-idf now

```
[09:49:38][I][bluetooth_proxy:250]: [0] [78:9C:85:19:4F:53] Connecting v3 with cache
[09:49:38][I][esp32_ble_client:064]: [0] [78:9C:85:19:4F:53] 0x00 Attempting BLE connection
[09:49:39][I][esp-idf:000]: W (477511) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x3e

[09:49:39][I][esp32_ble_client:141]: [0] [78:9C:85:19:4F:53] Connected
[09:49:39][I][esp-idf:000]: E (478183) BT_APPL: earlier enc was not done for same device


[09:49:39][I][esp-idf:000]: E (478185) BT_APPL: earlier enc was not done for same device


[09:49:39][I][esp32_ble_client:263]: [0] [78:9C:85:19:4F:53] auth complete. remote BD_ADDR: 789c85194f53
[09:49:39][E][esp32_ble_client:266]: [0] [78:9C:85:19:4F:53] auth fail reason = 0x55
[09:49:39][I][esp32_ble_client:263]: [1] [] auth complete. remote BD_ADDR: 789c85194f53
[09:49:39][E][esp32_ble_client:266]: [1] [] auth fail reason = 0x55
[09:49:39][I][esp32_ble_client:263]: [2] [] auth complete. remote BD_ADDR: 789c85194f53
[09:49:39][E][esp32_ble_client:266]: [2] [] auth fail reason = 0x55
[09:49:39][I][esp-idf:000]: E (478375) BT_BTM: BTM_GetSecurityFlags false
```

**Comment by jagheterfredrik** [comment]
Interesting, is ESP_GAP_BLE_SEC_REQ_EVT being called?

My device does not trigger it and thus encryption is never setup leading to char write returning GATT_INSUFFICIENT_ENCRYPTION. I assume your device is asking for it and thus complain that it is being re-setup.

Thanks a bunch for testing so quickly 👍

**Comment by jesserockz** [comment]
No this event is not received on the ESP

**Comment by jagheterfredrik** [comment]
Ok, so the `esp_ble_set_encryption(...)` is what ESP32-people call pairing. I'll rework this to implement the pair() function instead.

I assume your device is unpaired, unencrypted, “Bluetooth Smart” device

**Comment by jagheterfredrik** [comment]
relates to https://github.com/esphome/aioesphomeapi/pull/390

**Comment by jagheterfredrik** [comment]
OK! @jesserockz, I've tested the changes with my device and issuing a pair() command through aioesphomeapi works beautifully, e.g.:

```python
import aioesphomeapi
import asyncio
 
async def main():
    api = aioesphomeapi.APIClient("10.0.13.37", 6053, password=None)
    await api.connect(login=True)
 
    def _on_bluetooth_connection_state(
        connected: bool, mtu: int, error: int
    ) -> None:
        print(f"Connect cb! {connected=} {mtu=}")
 
    address = int("MY:DE:VI:CE:HE:RE".replace(":", ""), 16)
    await api.bluetooth_device_connect(
        address,
        _on_bluetooth_connection_state,
        timeout=10,
        has_cache=False,
        version=3,
        address_type=0,
    )
    await api.bluetooth_device_pair(address)
 
    def on_notify(int, data):
        print(data)
 
    await api.bluetooth_gatt_start_notify(address, 26, on_notify)
    print(await api.bluetooth_gatt_write(address, 29, bytes([3]), True))
    await asyncio.sleep(0.5)
    print(await api.bluetooth_gatt_write(address, 23, bytes(b'get sy v\r\n'), True))
 
    await asyncio.sleep(5)
    await api.bluetooth_device_disconnect(address)
 
loop = asyncio.get_event_loop()
loop.run_until_complete(main())
```

Without calling bluetooth_device_pair() the device returns GATT_INSUFFICIENT_ENCRYPTION.

**Comment by bdraco** [comment]
We need to bump the Bluetooth proxy version as well so hass can know it's available

**Comment by bdraco** [comment]
https://github.com/home-assistant/core/pull/88603/files#diff-972150b31eac2d4abbf3a9e71009b8ed20d32b4ae0c89253fc74d2248d5b6d1dR46

**Comment by jagheterfredrik** [comment]
I started https://github.com/jagheterfredrik/core/tree/bluetooth-proxy-pair but you're quick @bdraco, thank you! 🙏 I'll bump the version here

File: esphome/components/esp32_ble_client/ble_client_base.cpp [hunk 1]
```diff
@@ -72,6 +72,8 @@ void BLEClientBase::connect() {
   }
 }
 
+void BLEClientBase::pair() { esp_ble_set_encryption(this->remote_bda_, ESP_BLE_SEC_ENCRYPT); }
```
Comment (bdraco) [inline esphome/components/esp32_ble_client/ble_client_base.cpp]: esp_err_t esp_ble_set_encryption(esp_bd_addr_t bd_addr, esp_ble_sec_act_t sec_act)
Set a gap parameter value. Use this function to change the default GAP parameter values.

Parameters
bd_addr – [in] : the address of the peer device need to encryption
sec_act – [in] : This is the security action to indicate what kind of BLE security level is required for the BLE link if the BLE is supported
Returns
- ESP_OK : success

other : failed


We should check the return status and feed it back here

**Comment by jagheterfredrik** [comment]
I think for something useful we'd need to wait for the ESP_GAP_BLE_AUTH_CMPL_EVT event https://github.com/jagheterfredrik/esphome/blob/just-works/esphome/components/esp32_ble_client/ble_client_base.cpp#L256-L268

Would a new callback with auth status be the way to go?

I'll look at how the BlueZ Bleak backend handles it tonight.

**Comment by bdraco** [comment]
> Would a new callback with auth status be the way to go?
> 
> 

Yes I think so

**Comment by jagheterfredrik** [comment]
@jesserockz I’m quite pleased, let me know what you think

**Comment by jagheterfredrik** [comment]
Adding unpair aswell. Will get back to you when it’s in a good shape

**Comment by jagheterfredrik** [comment]
OK, PTAL. Happy to discuss naming.

File: esphome/components/bluetooth_proxy/bluetooth_connection.cpp [hunk 2]
```diff
@@ -158,6 +158,25 @@ bool BluetoothConnection::gattc_event_handler(esp_gattc_cb_event_t event, esp_ga
   return true;
 }
 
+void BluetoothConnection::gap_event_handler(esp_gap_ble_cb_event_t event, esp_ble_gap_cb_param_t *param) {
+  BLEClientBase::gap_event_handler(event, param);
+
+  switch (event) {
+    case ESP_GAP_BLE_AUTH_CMPL_EVT:
+      esp_bd_addr_t bd_addr;
+      memcpy(bd_addr, param->ble_security.auth_cmpl.bd_addr, sizeof(esp_bd_addr_t));
```
Comment (jesserockz) [inline esphome/components/bluetooth_proxy/bluetooth_connection.cpp]: This is not being used, so no need to copy it...
```suggestion
```

File: esphome/components/esp32_ble_client/ble_client_base.h [hunk 3]
```diff
@@ -71,6 +72,7 @@ class BLEClientBase : public espbt::ESPBTClient, public Component {
   void set_remote_addr_type(esp_ble_addr_type_t address_type) { this->remote_addr_type_ = address_type; }
   uint16_t get_conn_id() const { return this->conn_id_; }
   uint64_t get_address() const { return this->address_; }
+  bool get_paired() const { return this->paired_; }
```
Comment (jesserockz) [inline esphome/components/esp32_ble_client/ble_client_base.h]: Perhaps a better name?
```suggestion
  bool is_paired() const { return this->paired_; }
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
Mostly looks good, just a couple of requests below

File: esphome/components/esp32_ble_client/ble_client_base.cpp (hunk 1 above)
Comment (jesserockz) [reply esphome/components/esp32_ble_client/ble_client_base.cpp]: Dont forget this change request...

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/esp32_ble_client/ble_client_base.cpp (hunk 1 above)
Comment (jagheterfredrik) [reply esphome/components/esp32_ble_client/ble_client_base.cpp]: Ah we solved it by checking the auth completed callback instead, I'll resolve this comment

Comment (bdraco) [reply esphome/components/esp32_ble_client/ble_client_base.cpp]: If this errors, I think the callback will never fire.  I think we need to send the failure message right away if `esp_ble_set_encryption` returns anything other than ESP_OK

Comment (jagheterfredrik) [reply esphome/components/esp32_ble_client/ble_client_base.cpp]: Ok! Done-so 👍

**Comment by emontnemery** [comment]
There seems to be some issue with the latest version. I'm testing with two pairable devices, and Home Assistant often gets auth errors when trying to communicate with the devices.

It seems like the first device Home Assistant talks to works fine, whereas the 2nd device has issues. I've only tested a handful of times though, so I'm not sure if it's always the 2nd device which fails.

The error in the Home Assistant log (actually from bleak) say there's a problem with "insufficient authentication".

Note: The problem doesn't always occur, it happens maybe 3 out of 4 times when restarting Home Assistant.

In the ESPHome logs I see this:
```
[11:20:05][D][api.connection:920]: Home Assistant 2023.4.0.dev0 (192.168.0.5): Connected successfully
[11:20:09][I][bluetooth_proxy:254]: [0] [F0:94:0A:BD:3D:0A] Connecting v3 without cache
[11:20:09][D][esp32_ble_tracker:206]: Pausing scan to make connection...
[11:20:09][I][bluetooth_proxy:254]: [1] [ED:D2:90:2E:4C:69] Connecting v3 without cache
[11:20:09][I][esp32_ble_client:064]: [0] [F0:94:0A:BD:3D:0A] 0x01 Attempting BLE connection
[11:20:11][D][esp-idf:000]: W (106446) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x3e

[11:20:13][I][esp32_ble_client:196]: [0] [F0:94:0A:BD:3D:0A] Connected
[11:20:13][I][esp32_ble_client:064]: [1] [ED:D2:90:2E:4C:69] 0x01 Attempting BLE connection
[11:20:13][I][esp32_ble_client:261]: [0] [F0:94:0A:BD:3D:0A] auth complete. remote BD_ADDR: f0940abd3d0a
[11:20:13][I][esp32_ble_client:261]: [1] [ED:D2:90:2E:4C:69] auth complete. remote BD_ADDR: f0940abd3d0a
[11:20:13][I][esp32_ble_client:261]: [2] [] auth complete. remote BD_ADDR: f0940abd3d0a
[11:20:14][I][esp32_ble_client:081]: [0] [F0:94:0A:BD:3D:0A] Disconnecting.
[11:20:14][D][esp-idf:000]: W (109006) BT_HCI: hci cmd send: disconnect: hdl 0x0, rsn:0x13

[11:20:14][D][esp-idf:000]: W (109011) BT_APPL: gattc_conn_cb: if=3 st=0 id=3 rsn=0x16

[11:20:14][D][esp-idf:000]: W (109015) BT_APPL: gattc_conn_cb: if=4 st=0 id=4 rsn=0x16

[11:20:14][D][esp-idf:000]: W (109017) BT_APPL: gattc_conn_cb: if=5 st=0 id=5 rsn=0x16

[11:20:14][D][esp-idf:000]: W (109019) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x16

[11:20:14][I][esp32_ble_client:196]: [1] [ED:D2:90:2E:4C:69] Connected
[11:20:14][D][esp32_ble_tracker:238]: Starting scan...
[11:20:16][W][bluetooth_proxy.connection:101]: [1] [ED:D2:90:2E:4C:69] Error writing char/descriptor at handle 0x F, status=5
[11:20:16][W][bluetooth_proxy.connection:101]: [1] [ED:D2:90:2E:4C:69] Error writing char/descriptor at handle 0x12, status=5
[11:20:16][W][bluetooth_proxy.connection:101]: [1] [ED:D2:90:2E:4C:69] Error writing char/descriptor at handle 0x E, status=5
[11:20:16][W][bluetooth_proxy.connection:101]: [1] [ED:D2:90:2E:4C:69] Error writing char/descriptor at handle 0x E, status=5
[11:20:16][W][bluetooth_proxy.connection:101]: [1] [ED:D2:90:2E:4C:69] Error writing char/descriptor at handle 0x E, status=5
[11:20:16][W][bluetooth_proxy:217]: [1] [ED:D2:90:2E:4C:69] Connection already established
[11:20:17][W][bluetooth_proxy.connection:101]: [1] [ED:D2:90:2E:4C:69] Error writing char/descriptor at handle 0x F, status=5
[11:20:17][W][bluetooth_proxy.connection:101]: [1] [ED:D2:90:2E:4C:69] Error writing char/descriptor at handle 0x12, status=5
[11:20:17][W][bluetooth_proxy.connection:101]: [1] [ED:D2:90:2E:4C:69] Error writing char/descriptor at handle 0x E, status=5
[11:20:17][W][bluetooth_proxy.connection:101]: [1] [ED:D2:90:2E:4C:69] Error writing char/descriptor at handle 0x E, status=5
[11:20:17][W][bluetooth_proxy.connection:101]: [1] [ED:D2:90:2E:4C:69] Error writing char/descriptor at handle 0x E, status=5
[11:20:28][W][bluetooth_proxy:217]: [1] [ED:D2:90:2E:4C:69] Connection already established
```

Let me know what I can do to aid debugging.

**Comment by jagheterfredrik** [comment]
> There seems to be some issue with the latest version

So you mean that it was working before?

**Comment by emontnemery** [comment]
@jagheterfredrik I tested on an earlier commit, before the waiting for ACK after pairing was implemented, that had other issues, probably because there was no wait for pairing to complete?
By the way, this stands out to me, what's with the index `2`?
```
[11:20:13][I][esp32_ble_client:261]: [0] [F0:94:0A:BD:3D:0A] auth complete. remote BD_ADDR: f0940abd3d0a
[11:20:13][I][esp32_ble_client:261]: [1] [ED:D2:90:2E:4C:69] auth complete. remote BD_ADDR: f0940abd3d0a
[11:20:13][I][esp32_ble_client:261]: [2] [] auth complete. remote BD_ADDR: f0940abd3d0a
```

You can contact me on discord if you want to chat, I'm emontnemery#6618 there.

**Comment by jagheterfredrik** [comment]
I see. Would you mind bisecting?

**Comment by emontnemery** [comment]
I don't mind bisecting, but I think it gets a bit tricky since there's interdependencies between the Home Assistant Core, aioesphomeapi and esphome PRs?

If you can outline which combinations commits you'd like me to test or bisect, I'll do that though.

**Comment by jagheterfredrik** [comment]
esphome should suffice. Did you try to explicitly disconnect-> connect -> pair -> write?

**Comment by emontnemery** [comment]
> esphome should suffice

That's where I'm confused, if I just bisect the commits in this PR, I'll run into problems because aioesphome and home assistant are not compatible with ESPHome? Or do you have some suggestion for how to test ESPHome in isolation?

> Did you try to explicitly disconnect-> connect -> pair -> write

No, I did not try that.

**Comment by jagheterfredrik** [comment]
Latest aioesphomeapi should be backwards compatible. I’m not sure I follow how you are testing, is it a custom HA component?

I’d prefer to move this to a separate issue as it’s working on your other device. Insufficient auth could mean that the device is not happy with just-works pairing and requires a passcode pairing procedure which is not implemented and would require a lot of additional work.

**Comment by emontnemery** [comment]
> Latest aioesphomeapi should be backwards compatible

Is it really backwards compatible considering the protobuf definitions were changed several times, with new messages added and so on?

> I’m not sure I follow how you are testing, is it a custom HA component

I'm testing with a core integration which has support for a Bluetooth device which requires pairing: https://next.home-assistant.io/integrations/dormakaba_dkey/
My test involves two identical devices added to Home Assistant. Sometimes it's possible to talk to both of them, sometimes only one of them.

> I’d prefer to move this to a separate issue as it’s working on your other device

The pairing implemented by this PR seems to be flaky; *it sometimes works and it sometimes doesn't*, so I think that should be solved before this PR is merged?

Based on some more testing:
- It always seems to be successful to talk to the very first device Home Assistant contacts, that means the first config entry which is setup
- The 2nd config entry to be setup is not reliable
  - It's not a problem with the device themselves; the problem depends on setup order (verified by reordering config entries in `core.config_entries`)
  - The problem with the 2nd device seems to be triggered if Home Assistant is stopped, and then immediately started again.

Could you contact me on discord, I'm emontnemery#6618 there, I think it may be easier to explain how I'm testing etc. if it's still unclear?

**Comment by emontnemery** [comment]
I just took a another look at the error logs I posted when I first mentioned the issue, and it seems clear that ESPHome is mixing up the two connections:
```
[11:20:09][I][bluetooth_proxy:254]: [0] [F0:94:0A:BD:3D:0A] Connecting v3 without cache
[11:20:09][D][esp32_ble_tracker:206]: Pausing scan to make connection...
[11:20:09][I][bluetooth_proxy:254]: [1] [ED:D2:90:2E:4C:69] Connecting v3 without cache
[11:20:09][I][esp32_ble_client:064]: [0] [F0:94:0A:BD:3D:0A] 0x01 Attempting BLE connection
[11:20:11][D][esp-idf:000]: W (106446) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x3e

[11:20:13][I][esp32_ble_client:196]: [0] [F0:94:0A:BD:3D:0A] Connected
[11:20:13][I][esp32_ble_client:064]: [1] [ED:D2:90:2E:4C:69] 0x01 Attempting BLE connection
[11:20:13][I][esp32_ble_client:261]: [0] [F0:94:0A:BD:3D:0A] auth complete. remote BD_ADDR: f0940abd3d0a
[11:20:13][I][esp32_ble_client:261]: [1] [ED:D2:90:2E:4C:69] auth complete. remote BD_ADDR: f0940abd3d0a  <-- THIS IS THE ADDRESS OF THE 1ST DEVICE!!!!
[11:20:13][I][esp32_ble_client:261]: [2] [] auth complete. remote BD_ADDR: f0940abd3d0a <-- WHERE DOES INDEX 2 COME FROM, WE'RE TALKING TO 2 DEVICES, NOT 3
```

**Comment by emontnemery** [comment]
Here's a successful log, the difference is that the connections to the two devices do not happen in parallel:
```
[12:31:28][D][api.connection:920]: Home Assistant 2023.4.0.dev0 (192.168.0.5): Connected successfully
[12:31:35][I][bluetooth_proxy:254]: [0] [ED:D2:90:2E:4C:69] Connecting v3 without cache
[12:31:35][D][esp32_ble_tracker:206]: Pausing scan to make connection...
[12:31:35][I][esp32_ble_client:064]: [0] [ED:D2:90:2E:4C:69] 0x01 Attempting BLE connection
[12:31:37][I][esp32_ble_client:196]: [0] [ED:D2:90:2E:4C:69] Connected
[12:31:37][D][esp32_ble_tracker:238]: Starting scan...
[12:31:37][I][esp32_ble_client:261]: [0] [ED:D2:90:2E:4C:69] auth complete. remote BD_ADDR: edd2902e4c69
[12:31:37][I][esp32_ble_client:261]: [1] [] auth complete. remote BD_ADDR: edd2902e4c69
[12:31:37][I][esp32_ble_client:261]: [2] [] auth complete. remote BD_ADDR: edd2902e4c69
[12:31:39][I][esp32_ble_client:081]: [0] [ED:D2:90:2E:4C:69] Disconnecting.
[12:31:39][D][esp-idf:000]: W (57608670) BT_HCI: hci cmd send: disconnect: hdl 0x0, rsn:0x13

[12:31:39][D][esp-idf:000]: W (57608676) BT_APPL: gattc_conn_cb: if=3 st=0 id=3 rsn=0x16

[12:31:39][D][esp-idf:000]: W (57608678) BT_APPL: gattc_conn_cb: if=4 st=0 id=4 rsn=0x16

[12:31:39][D][esp-idf:000]: W (57608681) BT_APPL: gattc_conn_cb: if=5 st=0 id=5 rsn=0x16

[12:31:39][D][esp-idf:000]: W (57608688) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x16

[12:31:39][I][bluetooth_proxy:254]: [0] [F0:94:0A:BD:3D:0A] Connecting v3 without cache
[12:31:39][D][esp32_ble_tracker:206]: Pausing scan to make connection...
[12:31:39][I][esp32_ble_client:064]: [0] [F0:94:0A:BD:3D:0A] 0x01 Attempting BLE connection
[12:31:41][I][esp32_ble_client:196]: [0] [F0:94:0A:BD:3D:0A] Connected
[12:31:41][D][esp32_ble_tracker:238]: Starting scan...
[12:31:41][I][esp32_ble_client:261]: [0] [F0:94:0A:BD:3D:0A] auth complete. remote BD_ADDR: f0940abd3d0a
[12:31:41][I][esp32_ble_client:261]: [1] [] auth complete. remote BD_ADDR: f0940abd3d0a
[12:31:41][I][esp32_ble_client:261]: [2] [] auth complete. remote BD_ADDR: f0940abd3d0a
[12:31:42][I][esp32_ble_client:081]: [0] [F0:94:0A:BD:3D:0A] Disconnecting.
```

**Comment by emontnemery** [comment]
I got some help from @jesserockz, the problem is that the newly added code does not check if BLE events are for the correct connection slot. After adding that, it works 👍

**Comment by emontnemery** [comment]
This needs to be added:
```diff
diff --git a/esphome/components/bluetooth_proxy/bluetooth_connection.cpp b/esphome/components/bluetooth_proxy/bluetooth_connection.cpp
index f4504983..452f5fd8 100644
--- a/esphome/components/bluetooth_proxy/bluetooth_connection.cpp
+++ b/esphome/components/bluetooth_proxy/bluetooth_connection.cpp
@@ -163,6 +163,8 @@ void BluetoothConnection::gap_event_handler(esp_gap_ble_cb_event_t event, esp_bl

   switch (event) {
     case ESP_GAP_BLE_AUTH_CMPL_EVT:
+      if (memcmp(param->ble_security.auth_cmpl.bd_addr, this->remote_bda_, 6) != 0)
+        return;
       if (param->ble_security.auth_cmpl.success) {
         api::global_api_server->send_bluetooth_device_pairing(this->address_, true);
       } else {
diff --git a/esphome/components/esp32_ble_client/ble_client_base.cpp b/esphome/components/esp32_ble_client/ble_client_base.cpp
index 9654287f..9008e7bd 100644
--- a/esphome/components/esp32_ble_client/ble_client_base.cpp
+++ b/esphome/components/esp32_ble_client/ble_client_base.cpp
@@ -250,11 +250,15 @@ void BLEClientBase::gap_event_handler(esp_gap_ble_cb_event_t event, esp_ble_gap_
   switch (event) {
     // This event is sent by the server when it requests security
     case ESP_GAP_BLE_SEC_REQ_EVT:
+      if (memcmp(param->ble_security.ble_req.bd_addr, this->remote_bda_, 6) != 0)
+        return;
       ESP_LOGV(TAG, "[%d] [%s] ESP_GAP_BLE_SEC_REQ_EVT %x", this->connection_index_, this->address_str_.c_str(), event);
       esp_ble_gap_security_rsp(param->ble_security.ble_req.bd_addr, true);
       break;
     // This event is sent once authentication has completed
     case ESP_GAP_BLE_AUTH_CMPL_EVT:
+      if (memcmp(param->ble_security.auth_cmpl.bd_addr, this->remote_bda_, 6) != 0)
+        return;
       esp_bd_addr_t bd_addr;
       memcpy(bd_addr, param->ble_security.auth_cmpl.bd_addr, sizeof(esp_bd_addr_t));
       ESP_LOGI(TAG, "[%d] [%s] auth complete. remote BD_ADDR: %s", this->connection_index_, this->address_str_.c_str(),
```

**Comment by jagheterfredrik** [comment]
Thanks @emontnemery, nice debugging 👍

**Comment by emontnemery** [comment]
@jagheterfredrik do you have some example of BLE devices that require pairing? Is it the igrill device you need it for?

**Comment by jagheterfredrik** [comment]
I'm testing on a Wallbox Pulsar Plus (https://github.com/jagheterfredrik/wallbox-ble)

**Comment by emontnemery** [comment]
I've tested this a bit more now, with the address guards in place things seem to work fine every time 👍

**Review by bdraco — APPROVED** [review]

**Review by jesserockz — APPROVED** [review]

**Pull request merged**